### PR TITLE
Logfmt parser for parsing out additional fields from record msg field

### PR DESCRIPTION
**Pull request by minxuancao**
I implemented a parser by extending the k8Json parser so that we could parse out additional fields  from the msg field of each record. This assumes that the msg is in logfmt format(about logfmt: https://brandur.org/logfmt). The fields we want to parse out need to be specified in the Fields map in the schema of collect.json.  See below for an example.
```
    "Schemas": [
      {
        "PathMatcher": "/*(?:.+/)*(?P<file>.+\\..+)",
        "DataFormat": "text",
        "DateFormats": [
          "DD/MMM/YYYY:HH:mm:ss ZZZZ"
        ],
        "Meta": {
          "Tags": {
            "file": "{file}"
          }
          "Fields":{
         "field1": "field1"
         "field2": "field2"
         }
        }
      }
```  
Here is a sample use-case.
```
record: 
{"log":"type=hello level=debug","stream":"stderr","time":"2018-02-06T10:22:49.201168034Z"}
```
By specifying type, level in Fields, we could have type and level as additional fields and use in LQL by calling {vars:type} or {vars:level}
```
parsed result:
msg="type=hello level=debug", stream=stderr, time=2018-02-06T10:22:49.201168034Z, type=hello, level=debug
```
If msg contains a correctly formatted time field, then the parsed out field would replace the value in the time field of the msg.

```
record
{"log":"time=2018-01-01T10:22:49.71369883Z","stream":"stderr","time":"2018-02-06T10:22:49.201168034Z"}
```
```
parsed result
log="time=2018-01-01T10:22:49.71369883Z",stream=stderr, time=2018-01-01T10:22:49.71369883Z
```

I included tests to test the expected behavior.

One implementation detail to note is that after passing in the Field map to the parser, I take the keys in the map and save it in an array and delete all the keys in the map to avoid printing out duplicated fields in LQL. I would like to adjust the code if there is a better way of dealing with this.

### CODE REVIEW

File: pkg/scanner/parser/logfmt_parser.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	rec := model.NewRecord(*(*[]byte)(unsafe.Pointer(&r.Log)), r.Time)
+	rec.Fields = "stream=" + r.Stream
+
+	msg := make(Msg)
+	if err := logfmt.Unmarshal(rec.Data,&msg); err == nil{
```
Comment (dspasibenko) [inline pkg/scanner/parser/logfmt_parser.go]: run gofmt: `rec.Data,&msg` => `rec.Data, &msg`; ` nil{` => ` nil {` etc.

Comment (dspasibenko) [inline pkg/scanner/parser/logfmt_parser.go]: Question: so if a parsing error, we just ignore the line, right?

File: pkg/scanner/parser/logfmt_parser.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	lp.fields = make([]string,0,len(fieldMap))
+	for key,_ := range fieldMap{
+		lp.fields = append(lp.fields,key)
+		delete(fieldMap,key)
+	}
+
+	return lp, nil
+}
+
+func (m Msg) HandleLogfmt(key, val []byte) error {
```
Comment (dspasibenko) [inline pkg/scanner/parser/logfmt_parser.go]: Why the function has public visibility? Where is it used?

File: pkg/scanner/parser/logfmt_parser.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+)
+
+type (
+	// K8sJsonLogParser implements parser.Parser, for reading lines(strings with \n separation)
+	// from a text file, treating every line as a k8s json log message (fixed fields).
+	// The parser doesn't pay much attention to parsing dates, as they are presented as time.Time
+	// object in case of k8s log and in such a case JSON deserialization should take care of it.
+	// Additionally to message and date the parser also saves some additional metadata like 'pipe' (stdout/strerr),
+	// which is provided in k8s logs, the result of parsing is model.Record.
+	LogfmtParser struct {
```
[the remark this answers is not in the document]
Comment (minxuancao) [reply pkg/scanner/parser/logfmt_parser.go]: I don't think `logfmtParser` needs to be global. I think I made it global b/c I saw that `K8sJsonLogParser` is global. I have changed it. `K8sJsonLogParser` may need to be changed as well?

File: pkg/scanner/parser/logfmt_parser.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	}
+
+	// K8sJsonLogRec defines format of k8s json log line
+	LogfmtJsonLogRec struct {
+		Log    string    `json:"log"`
+		Stream string    `json:"stream"`
+		Time   time.Time `json:"time"`
+	}
+
+	Msg map[string]string
```
[the remark this answers is not in the document]
Comment (minxuancao) [reply pkg/scanner/parser/logfmt_parser.go]: I don't think we can change the visibility. The variable needs to be passed in a Unmarshall function.

File: pkg/scanner/parser/logfmt_parser.go (hunk 1 above)
Comment (minxuancao) [reply pkg/scanner/parser/logfmt_parser.go]: Yes. If there is a parsing error, we just skipping parsing the msg field of the record and return the record.

File: pkg/scanner/parser/logfmt_parser.go (hunk 2 above)
Comment (minxuancao) [reply pkg/scanner/parser/logfmt_parser.go]: it's used as a customHandler for logfmt: https://godoc.org/github.com/kr/logfmt#example-package--CustomHandler

**Comment by minxuancao** [comment]
The parser has been tested in the cmdline using LQL. They are sent as designed.

File: pkg/scanner/parser/logfmt_parser.go (hunk 1 above)
Comment (minxuancao) [reply pkg/scanner/parser/logfmt_parser.go]: addressed

File: pkg/scanner/parser/logfmt_parser.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+	"github.com/kr/logfmt"
+	"github.com/logrange/logrange/pkg/scanner/model"
+	"io"
+	"os"
+	"time"
+	"unsafe"
+)
+
+type (
+	// K8sJsonLogParser implements parser.Parser, for reading lines(strings with \n separation)
```
[the remark this answers is not in the document]
Comment (minxuancao) [reply pkg/scanner/parser/logfmt_parser.go]: changed

File: pkg/scanner/parser/parser.go [hunk 6]
```diff
@@ -71,6 +71,7 @@ const (
 	FmtPure   DataFormat = "pure"
 	FmtText   DataFormat = "text"
 	FmtK8Json DataFormat = "k8json"
+	FmtLogfmt DataFormat = "logfmt"
```
[the remark this answers is not in the document]
Comment (minxuancao) [reply pkg/scanner/parser/parser.go]: changed

File: pkg/scanner/schema.go [hunk 7]
```diff
@@ -120,7 +122,7 @@ func (sc *SchemaConfig) Check() error {
 	if err != nil {
 		return fmt.Errorf("PathMatcher=%v is invalid; %v", sc.PathMatcher, err)
 	}
-	if sc.DataFormat != parser.FmtK8Json && sc.DataFormat != parser.FmtText && sc.DataFormat != parser.FmtPure {
+	if sc.DataFormat != parser.FmtK8Json && sc.DataFormat != parser.FmtText && sc.DataFormat != parser.FmtPure && sc.DataFormat != parser.FmtLogfmt{
```
[the remark this answers is not in the document]
Comment (minxuancao) [reply pkg/scanner/schema.go]: changed

File: pkg/scanner/parser/parser.go [hunk 8]
```diff
@@ -85,8 +86,9 @@ func ToDataFormat(str string) (DataFormat, error) {
 	case FmtPure:
 	case FmtText:
 	case FmtK8Json:
+	case FmtLogfmt:
 	default:
-		err = fmt.Errorf("unknown date format \"%s\", expecting \"%s\", \"%s\" or \"%s\"", str, FmtPure, FmtText, FmtK8Json)
+		err = fmt.Errorf("unknown date format \"%s\", expecting \"%s\", \"%s\" or \"%s\" or \"%s\"", str, FmtPure, FmtText, FmtK8Json,FmtLogfmt)
```
[the remark this answers is not in the document]
Comment (minxuancao) [reply pkg/scanner/parser/parser.go]: changed

File: pkg/scanner/parser/logfmt_parser.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dspasibenko) [reply pkg/scanner/parser/logfmt_parser.go]: @minxuancao  yes, you can :). The fields must have the public visibility. `map` type has public visibility as well. It is always a good idea to write a test and check. For this particular case, you can make it.

[the remark this answers is not in the document]
Comment (minxuancao) [reply pkg/scanner/parser/logfmt_parser.go]: @dspasibenko sorry. Should have tried this first. I have tested and changed in the commit.

**Pull request merged**